nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/CaptureChangeMySQL.java

```diff
@@ -0,0 +1,928 @@
+        }
+
+        ComponentLog log = getLogger();
+        StateManager stateManager = context.getStateManager();
+
+        try {
+            outputEvents(currentSession, stateManager, log);
+        } catch (IOException ioe) {
+            try {
+                // Perform some processor-level "rollback", then rollback the session
+                currentBinlogFile = xactBinlogFile == null ? "" : xactBinlogFile;
+                currentBinlogPosition = xactBinlogPosition;
+                currentSequenceId.set(xactSequenceId);
+                inTransaction = false;
+                stop(stateManager);
+                queue.clear();
+                currentSession.rollback();
+            } catch (Exception e) {
+                // Not much we can recover from here
+            }
+            throw new ProcessException(ioe);
+        }
+    }
+
+    @OnStopped
+    public void onStopped(ProcessContext context) {
+        try {
+            stop(context.getStateManager());
+        } catch (CDCException ioe) {
+            throw new ProcessException(ioe);
+        }
+    }
+
+    /**
+     * Get a list of hosts from a NiFi property, e.g.
+     *
+     * @param hostsString A comma-separated list of hosts (host:port,host2:port2, etc.)
+     * @return List of InetSocketAddresses for the hosts
+     */
+    private List<InetSocketAddress> getHosts(String hostsString) {
+
+        if (hostsString == null) {
+            return null;
+        }
+        final List<String> hostsSplit = Arrays.asList(hostsString.split(","));
+        List<InetSocketAddress> hostsList = new ArrayList<>();
+
+        for (String item : hostsSplit) {
+            String[] addresses = item.split(":");
+            if (addresses.length != 2) {
+                throw new ArrayIndexOutOfBoundsException("Not in host:port format");
+            }
+
+            hostsList.add(new InetSocketAddress(addresses[0].trim(), Integer.parseInt(addresses[1].trim())));
+        }
+        return hostsList;
+    }
+
+    protected void connect(List<InetSocketAddress> hosts, String username, String password, boolean createEnrichmentConnection,
+                           String driverLocation, String driverName, long connectTimeout) throws IOException {
+
+        int connectionAttempts = 0;
+        final int numHosts = hosts.size();
+        InetSocketAddress connectedHost = null;
+
+        while (connectedHost == null && connectionAttempts < numHosts) {
+            if (binlogClient == null) {
+
+                connectedHost = hosts.get(currentHost);
+                binlogClient = createBinlogClient(connectedHost.getHostString(), connectedHost.getPort(), username, password);
+            }
+
+            if (eventListener == null) {
+                eventListener = createBinlogEventListener(binlogClient, queue);
+            }
+            eventListener.start();
+            binlogClient.registerEventListener(eventListener);
```

Comment:
Actually looking through the code, the BinaryLogClient connect(timeout) method will try to connect in a separate thread then await a CountDownLatch, throwing a TimeoutException if it doesn't work. Do we need a LifecycleListener for that? I was thinking alternatively we could catch exceptions in connect(), possibly building up a list of error messages received from the various connection attempts? If we were to use a LifecycleListener, perhaps we only need the onDisconnect() method, in case something bad happens while the processor is running? Otherwise the processor's event listener will just not get called, and no repair or notification would be performed.